cygnus-arcgis/src/main/java/com/telefonica/iot/cygnus/backends/arcgis/restutils/RestAuthentication.java

```diff
@@ -0,0 +1,200 @@
+
+            if (expirationElement == null) {
+                expirationElement = json.get(APP_RESPONSE_EXPIRES_TAG);
+                System.out.println("Expiration: " + expirationElement.getAsLong());
+            } else {
+                long expiration = expirationElement.getAsLong();
+                Instant expirationInstant = Instant.ofEpochMilli(expiration);
+                if (expirationInstant.isBefore(Instant.now())) {
+                    // token is expired.
+                    System.out.println("Token expired at " + expirationInstant);
+                    return null;
+                }
+            }
+        } else {
+            String errorDesc = "Invalid token response format.";
+            if (json.get("error") != null) {
+                errorDesc = json.get("error").toString();
+            }
+            throw new ArcgisException(errorDesc);
+        }
+
+        return result;
+    }
+
+    /**
+     * 
+     * @param tokenJson
+     * @param expirationTag
+     * @return
+     * @throws ArcgisException
+     */
+    public static LocalDateTime expirationFromJson(String tokenJson, String expirationTag)
+            throws ArcgisException {
+        LocalDateTime result = null;
+        if ("".equals(expirationTag)) {
+            expirationTag = ONLINE_RESPONSE_EXPIRES_TAG;
+        }
+
+        JsonParser parser = new JsonParser();
+        JsonObject json = parser.parse(tokenJson).getAsJsonObject();
+        JsonElement node = json.get(expirationTag);
+
+        if (node != null) {
+            JsonElement expirationElement = json.get(ONLINE_RESPONSE_EXPIRES_TAG);
+
+            if (expirationElement == null) {
+                expirationElement = json.get(APP_RESPONSE_EXPIRES_TAG);
+            }
+
+            long expiration = expirationElement.getAsLong();
+
+            result = LocalDateTime.ofInstant(Instant.ofEpochMilli(expiration),
+                    ZoneId.systemDefault());
+        } else {
+            String errorDesc = "Invalid token response format.";
+            if (json.get("error") != null) {
+                errorDesc = json.get("error").toString();
+            }
+            throw new ArcgisException(errorDesc);
+        }
+
+        return result;
+    }
+
+    /**
+     * Request a token for a non OAuth token authentication.
+     * 
+     * @param cred
+     * @param tokenGenUrl
+     * @param referer
+     * @param expirationMins
+     * @return token
+     * @throws ArcGisException
+     */
+    public static Credential createUserToken(String user, String password, URL tokenGenUrl,
+            String referer, Integer expirationMins) throws ArcgisException {
+        String tokenJSON = null;
+        try {
+            Map<String, String> bodyParams = new LinkedHashMap<String, String>();
+            bodyParams.put("username", user);
+            bodyParams.put("password", password);
+            bodyParams.put("referer", referer);
+            bodyParams.put("username", user);
```

Comment:
It would be better to declare this Strings as constants.

Reply:
Agree.